hypothesis-python/src/hypothesis/control.py

```diff
@@ -139,12 +139,15 @@ def __init__(self, data, *, is_final=False, close_on_capture=True):
         self.known_object_printers = defaultdict(list)
 
     def record_call(self, obj, func, args, kwargs):
-        name = get_pretty_function_description(func)
-        self.known_object_printers[IDKey(obj)].append(
-            lambda obj, p, cycle: p.maybe_repr_known_object_as_call(
-                obj, cycle, name, args, kwargs
+        # We only make use of this for failing examples, so let's avoid the
+        # get_pretty_function_description overhead in the typical case.
+        if should_note():
+            name = get_pretty_function_description(func)
+            self.known_object_printers[IDKey(obj)].append(
```

Comment:
I think in this case we do actually want to record the call for e.g. observability mode, but we can still defer the (minor?) cost of `get_pretty_function_description()`:

```python
        self.known_object_printers[IDKey(obj)].append(
            lambda obj, p, cycle, *, _func=func: p.maybe_repr_known_object_as_call(
                obj, cycle, get_pretty_function_description(_func), args, kwargs
            )
        )
```

Reply:
I almost did this, but I wasn't sure we wanted to be holding a reference to `func` which might plausibly be eg a fresh function for every example/strategy. (if you think that's unlikely to cause issues then I'm happy with this alternative).

`get_pretty_function_description` is usually cheap...except when it hits `extract_lambda_source` on a new function for every example, which is the case for `FixedKeysDictStrategy` 😅. We can change this by lifting this to a function:

https://github.com/HypothesisWorks/hypothesis/blob/ba21999d34f0ad251d577b48d6d14a40dfc4c204/hypothesis-python/src/hypothesis/strategies/_internal/collections.py#L321

but I have low confidence in maintaining that going forward given the opaqueness of the lambda vs function <-> pretty printer relationship.

Reply:
good call on the binding semantics with `_func=func`, btw - I certainly would have missed this myself!

Reply:
The lifetime of the known-objects mapping is only that specific example, so I think it's basically fine.

Reply:
ah, right.